src/DurableTask.Netherite/StorageLayer/Faster/FasterKV.cs

```diff
@@ -950,187 +953,205 @@ public override ValueTask RemoveKeys(IEnumerable<TrackedObjectKey> keys)
+
+                            this.partition.Assert(orchestrationState == null || orchestrationState.OrchestrationInstance.InstanceId == instanceState.InstanceId, "wrong instance id");
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            this.partition.EventDetailTracer?.TraceEventProcessingDetail($"query {queryId} attempt {attempt:o} cancelled");
+                            goto done;
+                        }
+                        catch (Exception exception) when (this.terminationToken.IsCancellationRequested && !Utils.IsFatal(exception))
+                        {
+                            this.partition.EventDetailTracer?.TraceEventProcessingDetail($"query {queryId} attempt {attempt:o} cancelled due to partition termination");
+                            cancellationTokenSource.Cancel();
+                            goto done;
+                        }
+                        catch (Exception e)
                         {
-                            if (matched >= pageLimit.Value)
+                            this.partition.EventTraceHelper.TraceEventProcessingWarning($"query {queryId} attempt {attempt:o} enumeration failed with exception {e}");
+                            cancellationTokenSource.Cancel();
+                            goto done;
+                        }
+
+                        if (orchestrationState != null && instanceQuery.Matches(orchestrationState))
+                        {
+                            matched++;
+                            this.partition.EventDetailTracer?.TraceEventProcessingDetail($"match instance {queryRequests.Current}");
+                            yield return (position, orchestrationState);
+
+                            if (pageLimit.HasValue)
                             {
-                                cancellationTokenSource.Cancel();
-                                break;
+                                if (matched >= pageLimit.Value)
+                                {
+                                    cancellationTokenSource.Cancel();
+                                    break;
+                                }
+                            }
+                            else
+                            {
+                                pageCapacity.Release();
                             }
                         }
                         else
                         {
-                            leftToFill.Release();
+                            pageCapacity.Release();
                         }
                     }
-                    else
-                    {
-                        leftToFill.Release();
-                    }
                 }
+            pageDone:
+                yield return (position, null);
+                ReportProgress("completed-page");
+            done:
+                yield break;
+            }
+            finally
```

Comment:
Is there a specific reason you added a try/finally wrapper? I suppose it is for handling additional places where exceptions could be thrown that are not already caught?

Reply:
That's precisely right. I just want to make sure there's no usages of the semaphore after it's been disposed and, to do that, I need to make sure the producer task is awaited before we exit this method (as after we exit, the semaphore will be disposed)